py3status/modules/air_quality.py

```diff
@@ -0,0 +1,190 @@
+
+Example:
+```
+air_quality {
+    auth_token = 'demo'
+    location = 'Shanghai'
+    format = 'Shanghai: {aqi} {category}'
+}
+```
+@author beetleman
+@license BSD
+"""
+
+try:
+    # python 3
+    from urllib.error import URLError
+    from urllib.request import urlopen, Request
+    from urllib.parse import urlencode
+except ImportError:
+    # python 2
+    from urllib import urlencode
+    from urllib2 import URLError
+    from urllib2 import urlopen, Request
+import json
+
+AQI = (
+    (0, 50, 'icon_good'),
+    (51, 100, 'icon_moderate'),
+    (101, 150, 'icon_unhealthy_sensitive'),
+    (151, 200, 'icon_unhealthy'),
+    (201, 300, 'icon_very_unhealthy'),
+    (300, None, 'icon_hazardous')
+)
+URL = 'http://api.waqi.info'
+STRING_ERROR = 'air_quality: unknown error'
+
+
+def get_in(coll, path, default=None):
+    first = path[0]
+    rest = path[1:len(path)]
+    if len(path) == 1:
+        return coll.get(first, default)
+    return get_in(coll.get(first, default), rest, default)
+
+
+class Py3status:
+    auth_token = 'demo'
+    cache_timeout = 3600
+    color_good = '#009966'
+    color_hazardous = '#7E0023'
+    color_moderate = '#FFDE33'
+    color_unhealthy = '#CC0033'
+    color_unhealthy_sensitive = '#FF9933'
+    color_unknown = '#FFFFFF'
+    color_very_unhealthy = '#660099'
+    format = 'AQI: {aqi} {category}'
+    icon_good = 'Good'
+    icon_hazardous = 'Hazardous'
+    icon_moderate = 'Moderate'
+    icon_unhealthy = 'Unhealthy'
+    icon_unhealthy_sensitive = 'Unhealthy for Sensitive Groups'
+    icon_unknown = 'Unknown'
+    icon_very_unhealthy = 'Very Unhealthy'
+    location = 'Shanghai'
+
+    def _call_api(self):
+        kwargs = {
+            'token': self.auth_token
+        }
+        url_part = 'feed/{}'.format(self.location)
+
+        url = '{base_url}/{url_part}/?{query}'.format(
+            base_url=URL,
+            url_part=url_part,
+            query=urlencode(kwargs)
+        )
+
+        req = Request(url)
+        try:
+            resp = urlopen(req)
+            return json.loads(resp.read().decode('utf8'))
+        except URLError:
+            return {}
+
+    def _key(self, data):
+        if data.get('status') == 'ok':
+            aqi = get_in(data, ['data', 'aqi'], -1)
+            if type(aqi) is not int:
+                return 'unknown'
+
+            for start, end, key in AQI:
+                if aqi >= start and aqi <= end:
+                    return key
+
+        return 'unknown'
+
+    def _full_text(self, data):
+        key = self._key(data)
+        category = getattr(self, key, self.icon_unknown)
+        iaqi = get_in(data, ['data', 'iaqi'], {})
+        aqi = get_in(data, ['data', 'aqi'])
+        return self.format.format(
+            location=self.location,
+            category=category,
+            aqi=aqi,
+            **{k: _v.get('v') for k, _v in iaqi.items()}
+        )
+
+    def _color(self, data):
+        key = self._key(data)
```

Comment:
you are calling this both in `_color()` and `_full_text()` IMHO this makes a) the code harder to follow  b) we are doing more work than we need.  We should just get the key once and pass it to the functions.